### PR TITLE
Ensure no ancestor commit of parent branch is included into the branch's unique range of commits

### LINKED ISSUE #797: In case of red edge, no ancestor commit of parent branch should be included into the branch's unique range of commits
As for now, only reflogs of other branches are checked... but not logs, which is a shame.

See https://github.com/VirtusLab/git-machete-intellij-plugin/issues/1496

### PR DESCRIPTION
**Pull request by PawelLipski**

### CODE REVIEW

File: git_machete/git_operations.py [hunk 1]
```diff
@@ -774,15 +774,23 @@ def is_ancestor_or_equal(
             earlier_revision: AnyRevision,
             later_revision: AnyRevision,
     ) -> bool:
+        earlier_hash = self.get_commit_hash_by_revision(earlier_revision)
+        return self.get_merge_base(earlier_revision, later_revision) == earlier_hash
+
+    def get_merge_base(
```
Comment (amalota) [inline git_machete/git_operations.py:786]: @PawelLipski, it is confusing to have almost the same method name as the one method above -> `__get_merge_base` and call it inside this function. Maybe change the name of one of the functions (__get_merge_base?) or remove one and consolidate the code in the other, wdyt?

Comment (PawelLipski) [reply git_machete/git_operations.py:786]: Whoops right... reworked a bit

File: git_machete/client.py [hunk 2]
```diff
@@ -943,10 +943,8 @@ def fork_point_hash(branch_: LocalBranchShortName) -> Optional[FullCommitHash]:
                 try:
                     # We're always using fork point overrides, even when status
                     # is launched from discover().
-                    fork_point_hash_cached[branch_], fork_point_branches_cached[branch_] = self.__fork_point_and_containing_branch_pairs(
-                        branch_,
-                        use_overrides=True,
-                        opt_no_detect_squash_merges=opt_no_detect_squash_merges)
```
Comment (PawelLipski) [inline git_machete/client.py:949]: Unrel; the param was unused

File: git_machete/client.py [hunk 3]
```diff
@@ -1184,7 +1181,7 @@ def __fork_point_and_containing_branch_pairs(self,
             if overridden_fp_hash:
                 if upstream and \
                         self.__git.is_ancestor_or_equal(upstream.full_name(), branch.full_name()) and \
-                        not self.__git.is_ancestor_or_equal(upstream.full_name(), overridden_fp_hash.full_name()):
```
Comment (PawelLipski) [inline git_machete/client.py:1187]: Unrel: `.full_name()` is redundant on `FullCommitHash`

File: git_machete/client.py [hunk 4]
```diff
@@ -1225,8 +1224,13 @@ def __fork_point_and_containing_branch_pairs(self,
                     f"{upstream} is descendant of its upstream {branch}, but inferred fork point commit {fp_hash} "
                     f"is NOT a descendant of {upstream}; falling back to {upstream} as fork point")
                 return self.__git.get_commit_hash_by_revision(upstream), []
+            elif upstream and \
+                    not self.__git.is_ancestor_or_equal(upstream.full_name(), branch.full_name()) and \
+                    self.__git.is_ancestor_or_equal(fp_hash, upstream.full_name()):
+
+                common_ancestor_hash = self.__git.get_merge_base(upstream.full_name(), branch.full_name())
```
Comment (PawelLipski) [inline git_machete/client.py:1236]: Lemme put some logs in the newly-added cases...

**Review by amalota — APPROVED** [review]

**Pull request merged**